src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala

```diff
@@ -32,142 +30,48 @@ import scala.collection.mutable
-   *
-   * @param stopWordsA  Array of words that are filtered out.
-   */
-  class UniqueWordExtractor(stopWordsA: Array[String])
-    extends FlatMapFunction[(String, String), (String, Int) ] {
-
-    val stopWords = new mutable.HashSet[String]
-    val uniqueWords = new mutable.HashSet[String]
-    // initalize pattern to match words
-    val wordPattern: Pattern = Pattern.compile("(\\p{Alpha})+")
-
-    // initialize set of stop words
-    for(sw <- stopWordsA) {
-      this.stopWords.add(sw)
-    }
-
-    override def flatMap(t: (String, String), out: Collector[(String, Int)]): Unit = {
-      // clear unique words
-      uniqueWords.clear()
-
-      // split mail along whitespaces
-      val tokens = new StringTokenizer(t._2)
-      // for each word candidate
-      while(tokens.hasMoreTokens) {
-        // normalize word to lower case
-        val word = tokens.nextToken.toLowerCase
-        if (!stopWords.contains(word) && wordPattern.matcher(word).matches) {
-          // word candiate is not a stop word and matches the word pattern
-          uniqueWords.add(word)
-        }
-      }
-
-      // emit all words that occurred at least once
-      for(w <- uniqueWords) {
-        out.collect( (w, 1) )
-      }
-    }
+object MailTFIDF {
+  def main(args: Array[String]) {
+    val params = ParameterTool.fromArgs(args)
+    val input = params.getRequired("input")
+    val odir = params.get("odir", "/tmp")
+
+    // set up the execution environment
+    val env = ExecutionEnvironment.getExecutionEnvironment
+
+    val stopWords = Set(
+      "the", "i", "a", "an", "at", "are", "am", "for", "and", "or", "is", "there", "it", "this",
+      "that", "on", "was", "by", "of", "to", "in", "to", "not", "be", "with", "you",
+      "have", "as", "can")
+
+    val wordPattern = "(\\p{Alpha})+".r
+
+    val isWord = (s : String) => (!stopWords.contains(s) && wordPattern.findFirstIn(s).isDefined)
+
+    val mails = env.readCsvFile[(String, String)](
+      input,
+      lineDelimiter = "##//##",
+      fieldDelimiter = "#|#",
+      includedFields = Array(0,4)
+    // ) map (m => (m._1, m._2.toLowerCase.split("\\W+").collect{case s if isWord(s) => s})) first (1000)
+    ) map (m => (m._1, m._2.toLowerCase.split("\\W+").filter(s => (isWord(s))))) first (1000) // TODO : remove first (1000) i was running on a vm
+
+    val cnt = mails.count
+    // val tfs = mails flatMap (m => m._2.groupBy(w => w).mapValues(_.size).map(e => (m._1, e._1, e._2)))
+    val tfs = mails flatMap (m => m._2.groupBy(w => w).map(e => (m._1, e._1, e._2.size)))
+    val dfs = (mails flatMap (m => m._2.toSet)) map (m => (m, 1)) groupBy(0) reduce ((l, r) => (l._1, l._2 + r._2))
+    val tfidfs = tfs.join(dfs).where(1).equalTo(0) { (l, r) => (l._1, l._2, l._3 * (cnt.toDouble / r._2) ) }
+
+    // mails.map(m => (m._1, m._2.mkString(" "))).writeAsText(s"file://${odir}/MAILS.csv", WriteMode.OVERWRITE)
+    tfs.writeAsText(s"file://${odir}/TFs.csv", WriteMode.OVERWRITE)
```

Comment:
Please do not change the behavior of the job and only print the final result. 
This job is used as a reference implementation for the Flink training. We tried that all reference implementations follow a common pattern and I do not want to update the exercise descriptions. Thanks